src/sentry/api/endpoints/user_notification_fine_tuning.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import absolute_import
+
+import six
+from collections import defaultdict
+
+from django.db import transaction
+from rest_framework import status
+from rest_framework.response import Response
+
+from sentry.api.bases.user import UserEndpoint
+from sentry.api.serializers import serialize, Serializer
+from sentry.models import OrganizationMember, OrganizationMemberTeam, OrganizationStatus, ProjectTeam, UserOption, UserEmail
+
+
+KEY_MAP = {
+    'alerts': {
+        'key': 'mail:alert',
+        'type': int,
+    },
+
+    'workflow': {
+        'key': 'workflow:notifications',
+        'type': '',
+    },
+
+    'deploy': {
+        'key': 'deploy-emails',
+        'type': '',
+    },
+
+    'reports': {
+        'key': 'reports:disabled-organizations',
+        'type': '',
+    },
+
+    'email': {
+        'key': 'mail:email',
+        'type': '',
+    }
+}
+
+
+class UserNotificationsSerializer(Serializer):
+    def get_attrs(self, item_list, user, *args, **kwargs):
+        notification_type = kwargs['notification_type']
+        filter_args = {}
+
+        if notification_type in ['alerts', 'workflow', 'email']:
+            filter_args['project__isnull'] = False
+        elif notification_type == 'deploy':
+            filter_args['organization__isnull'] = False
+
+        data = list(UserOption.objects.filter(
+            key=KEY_MAP[notification_type]['key'],
+            user__in=item_list,
+            **filter_args
+        ).select_related('user', 'project', 'organization'))
+
+        results = defaultdict(list)
+
+        for uo in data:
+            results[uo.user].append(uo)
+
+        return results
+
+    def serialize(self, obj, attrs, user, *args, **kwargs):
+        notification_type = kwargs['notification_type']
+        data = {}
+
+        for uo in attrs:
+            if uo.project is not None:
+                data[uo.project.id] = uo.value
+            elif uo.organization is not None:
+                data[uo.organization.id] = uo.value
+            elif notification_type == 'reports':
+                # UserOption for key=reports:disabled-organizations saves a list of orgIds
+                # that should not receive reports
+                for org_id in uo.value:
+                    data[org_id] = 0
+        return data
+
+
+class UserNotificationFineTuningEndpoint(UserEndpoint):
+    def get(self, request, user, notification_type):
+        if notification_type not in KEY_MAP:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        notifications = UserNotificationsSerializer()
+
+        serialized = serialize(
+            user,
+            request.user,
+            notifications,
+            notification_type=notification_type
+        )
+        return Response(serialized)
+
+    def put(self, request, user, notification_type):
+        if notification_type not in KEY_MAP:
```

Comment:
i'm kind of confused about what the payload is for this endpoint. is it project or org id as keys and whether it's enabled or disabled as the value? i think it'd be helpful to heavily comment this, including what the payload should look like, how you know which type of id it is, etc (i'm guessing by what type of notification it is?)

Reply:
Hmm yeah this endpoint is kind of overloaded.